common/placer_heap.cc

```diff
@@ -544,28 +523,49 @@ class HeAPPlacer
                 bool placed = false;
                 int attempt_count = 0;
                 while (!placed) {
-                    if (!available_bels.count(ci->type) || available_bels.at(ci->type).empty())
-                        log_error("Unable to place cell '%s', no Bels remaining of type '%s'\n", ci->name.c_str(ctx),
-                                  ci->type.c_str(ctx));
+                    if (!available_bels.count(ci->type) || available_bels.at(ci->type).empty()) {
+                        log_error("Unable to place cell '%s', no BELs remaining to implement cell type '%s'\n",
+                                ci->name.c_str(ctx),
+                                ci->type.c_str(ctx));
+                    }
                     ++attempt_count;
-                    if (attempt_count > 25000)
+                    if (attempt_count > 25000) {
                         log_error("Unable to find a placement location for cell '%s'\n", ci->name.c_str(ctx));
-                    BelId bel = available_bels.at(ci->type).back();
-                    available_bels.at(ci->type).pop_back();
+                    }
+
+                    // Find an unused BEL from bels_for_cell_type.
+                    auto &bels_for_cell_type = available_bels.at(ci->type);
+                    BelId bel;
+                    while(true) {
+                        BelId candidate_bel = bels_for_cell_type.back();
```

Comment:
is there some error handling needed for the case that `bels_for_cell_type` might become empty during the loop?

Reply:
Ya, fixed.